lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -786,15 +801,53 @@ listPools spl =
             apparentPerformance
 
 joinStakePool
-    :: forall n k.
+    :: forall ctx s t n k.
         ( PaymentAddress n k
+        , Buildable (ErrValidateSelection t)
+        , s ~ SeqState n k
         , k ~ ShelleyKey
+        , HardDerivation k
+        , ctx ~ ApiLayer s t k
         )
-    => ApiT PoolId
+    => ctx
+    -> StakePoolLayer IO
+    -> ApiT PoolId
     -> ApiT WalletId
     -> ApiWalletPassphrase
     -> Handler (ApiTransaction n)
-joinStakePool _ _ _ = throwError err501
+joinStakePool ctx spl (ApiT poolId) (ApiT wid) passwd = do
+    (_, walMeta, _) <- liftHandler $ withWorkerCtx ctx wid throwE $
+        \wrk -> W.readWallet wrk wid
+    when (walMeta ^. #delegation == W.Delegating poolId) $
+        liftHandler $ throwE (ErrCreateCertPoolAlreadyJoined poolId)
+
+    allPools <- listPools spl
+    case L.find (\(ApiStakePool pId _ _) -> pId == ApiT poolId) allPools of
+        Nothing ->
+            liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
+        Just _ -> do
+            let (ApiWalletPassphrase (ApiT pwd)) = passwd
```

Comment:
**[suggestion: readability]**

I'd suggest to move the guard in a dedicated function (possibly declared in the where clause) to have something which is more consistent with the style adopted so far:

```hs
whenPoolIsUnknown poolId  $ liftHandler $ throwE (ErrCreateCertNoSuchPool poolId)
```

**Why**

This keeps the `joinStakePool` function more readable by moving some lower-level concerns into a separate function. Then, `joinStakePool` reads more like a sentence.

Reply:
done